src/hotspot/share/runtime/thread.cpp

```diff
@@ -4595,6 +4584,13 @@ void Threads::remove(JavaThread* p, bool is_daemon) {
   // Reclaim the ObjectMonitors from the om_in_use_list and om_free_list of the moribund thread.
   ObjectSynchronizer::om_flush(p);
 
+  // We must flush any deferred card marks and other various GC barrier
+  // related buffers (e.g. G1 SATB buffer and G1 dirty card queue buffer)
+  // before removing a thread from the list of active threads.
+  // This must be done after ObjectSynchronizer::om_flush(), as GC barriers
+  // are used in om_flush().
+  BarrierSet::barrier_set()->on_thread_detach(p);
+
```

Comment:
One last question, that doesn't require a comment why it's here, but why was this moved?

Reply:
We had to move that code to make om_flush() happy.
om_flush() accesses the (weak) oops in the monitor list
after the thread is off the threads list so the barrier change
had to move. It will move back in part3.

Reply:
Sorry, I should have been more clear here.
We had to move that code after the om_flush() call because om_flush()
accesses the (weak) oops in the monitor list after the original location
of the code. The om_flush() call is made in the code that removes the
thread from the threads list, but the removal happens after the call to
om_flush().